README.md
# sdx-console

[![Build Status](https://travis-ci.org/ONSdigital/sdx-console.svg?branch=master)](https://travis-ci.org/ONSdigital/sdx-console)

The Survey Data Exchange (SDX) Console is a component of the Office of National Statistics (ONS) SDX project. It will be used to test different microservices in SDX individually and to Reprocess Transactions for Operational Support.

Should only be deployed on-demand as required in an environment.

## Installation
This application presently installs required packages from requirements files:
- `requirements.txt`: packages for the application, with hashes for all packages: see https://pypi.org/project/hashin/ 
- `test-requirements.txt`: packages for testing and linting

It's also best to use `pyenv` and `pyenv-virtualenv`, to build in a virtual environment with the currently recommended version of Python.  To install these, see:
- https://github.com/pyenv/pyenv
- https://github.com/pyenv/pyenv-virtualenv
- (Note that the homebrew version of `pyenv` is easiest to install, but can lag behind the latest release of Python.)

To install, set your Python version using pyenv or similar, and run:
```
$ make build
```

To test, first run `make build` as above, then run:
```
$ make test
```

Alternatively, the `sdx-compose` repo allows you to spin up a local test environment for the whole ``sdx-`` suite of services. For more information, refer to the README in `sdx-compose`.

## Usage

To use the console you will need to start all services
Console requires a PostgreSQL database to the appropriate sdx-compose branch should be used

## Authentication

This application uses Flask-Security for login and Flask-Admin for user management

You will need to have an SDX-Console account set up by an administrator to access sdx-console while deployed

Locally one account will be created on launch: 'admin'. This user password is default 'admin'. This user can create developer users.

The login page is located at '/Login'. You can log out at '/Logout'

If you the administrator role you can access the user management AI (Flask-Admin) at '/admin'

## UI

 * `/store` -  allows searching of the PostgreSQL and reprocessing of transactions

## Configuration

Some of important environment variables available for configuration are listed below:

| Environment Variable    | Example                               | Description
|-------------------------|---------------------------------------|----------------
| SDX_STORE_URL           | `http://sdx-store:5000/`              | URL of the ``sdx-store`` service
| CONSOLE_LOGIN_TIMEOUT   | `10`                                  | Number of minutes to user timeout
| CONSOLE_INITIAL_ADMIN_PASSWORD| `PA$$W0RD` | Initial password used for user with admin rights



CHANGELOG.md
### Unreleased
  - Add low carbon
  - Updated UKIS
  - Add MES example
  - Add EPE examples

### 2.2.0 2020-06-09
  - Add construction example for formtype 0002
  - Remove Cloudfoundry deployment files

### 2.1.0 2020-05-21
  - Add example submissions for E-commerce formtype 0001
  - Add example submissions for Vacancies
  - Add construction example survey
  - Add covid-19 example survey
  - Update QBS example survey with new comment field
  - Update Vacancies survey examples
  - Updated packages

### 2.0.1 2019-11-08
  - Add python 3.8 to travis build
  - Update sdc-cryptography to 0.4.0
  - Update sdc-rabbit to 1.7.0 to fix reconnection issues
  - Update various other dependencies

### 2.0.0 2019-06-20
  - Added SEFT tab
  - Updated sdc-rabbit to allow use with python 3.7
  - modified 144.0002 data (UKIS NI) to have elements 0440 and 2671
  - Log version number on startup, at info level
  - Add configurable log level
  - Add change log
  - Fix [#36](https://github.com/ONSdigital/sdx-console/issues/36) crash/hang when FTP takes too long to respond
  - Fix [#37](https://github.com/ONSdigital/sdx-console/issues/37) incorrect "Enable empty FTP" setting - now off by default
  - Add QBS template
  - Add all environment variables to README
  - Removed old functionality
  - Added heartbeat
  - Added unittests
  - Added new decrypt functionality with unit tests
  - Added Authentication with flask_security
  - Added Account Management with flask_admin
  - Added database.py for handling PostgreSQL
  - Added Search functionality for sdx-store at /store
  - Adding sdx-common functionality
  - Updating logger format using sdx-common
  - Adding reprocessing functionality
  - Updating reprocessing functionality, removing queuepublisher and encrypter
  - Added Pagination
  - Added auditable logging
  - Added searching by valid/invalid
  - Added protection on "reprocess all" button
  - Add ability to add multiple user accounts dynamically
  - Improved login/out navigation
  - Fix login functionality with compose
  - Remove Decrypt Tab
  - Add multiple flask blueprints to improve readability
  - Add sdc patterns and common css/js
  - Improve UX throughout
  - Add submitter functionality
  - Remove sdx-common logging
  - Add validation on tx_id search to force a valid UUID string to be entered
  - FTP files write to memory instead of writing temporary files
  - Add pagination in store data
  - Create database tables in development mode
  - Remove /clear endpoint
  - Use pytest and codecov
  - Add tests to travis build
  - Updated unit tests to use an in memory database
  - Remove second rabbit host for aws
  - Add MBS json 167, 201, 255
  - Added D-Trades json
  - Disable browser caching of sample submission data

### 1.7.0 2017-11-30
  - Deploying new console

### 1.6.0 2017-11-21
  - Remove sdx-common logging and move logging code back into the repo
  - Add Cloudfoundry deployment files

### 1.5.1 2017-10-25
  - Add KID hash to the encrypter

### 1.5.0 2017-09-14
  - Remove sdx-common git clone in Dockerfile

### 1.4.0 2017-08-04
  - Add MakeFile

### 1.3.0 2017-07-25
  - Update example json for EQ Survey refresh
  - Change all instances of ADD to COPY in Dockerfile

### 1.2.0 2017-07-10
  - Adding tx_id to the message headers
  - Log version number on startup, at info level
  - Add configurable log level
  - Add change log
  - Fix [#36](https://github.com/ONSdigital/sdx-console/issues/36) crash/hang when FTP takes too long to respond
  - Fix [#37](https://github.com/ONSdigital/sdx-console/issues/37) incorrect "Enable empty FTP" setting - now off by default
  - Add QBS template
  - Add all environment variables to README
  - Remove secure data from logging messages
  - Add license
  - Adding sdx-common functionality
  - Updating logger format using sdx-common
  - Update and pin version of sdx-common to 0.7.0

### 1.0.0 2016-08-16
  - Initial release


console/views.py
import ast
import json
import logging
import math

from datetime import datetime
from flask import jsonify
from flask import redirect
from flask import render_template
from flask import request
from flask import url_for
import flask_security
import requests
from sqlalchemy import func
from sqlalchemy.exc import DataError, SQLAlchemyError
from structlog import wrap_logger

from . import app, create_dev_user, settings
from .database import db_session
from .forms import NewUserForm
from .helpers.exceptions import ClientError, ResponseError, ServiceError, UserCreationError, UserExistsError
from .models import SurveyResponse

logger = wrap_logger(logging.getLogger(__name__))


@app.errorhandler(ResponseError)
def handle_invalid_usage(error):
    json_error = {"message": error.message, "status_code": error.status_code}
    return jsonify(json_error)


@app.route('/healthcheck', methods=['GET'])
def healthcheck():
    return jsonify({'status': 'OK'})


@app.route('/', methods=['GET'])
def home():
    return render_template('home.html',
                           current_user=flask_security.core.current_user,
                           request=request)


def send_data(logger, url, data=None, json=None, request_type="POST"):
    response_logger = logger.bind(url=url)
    try:
        if request_type == "POST":
            response_logger.info("Sending POST request")
            if data:
                r = requests.post(url, data=data)
            elif json:
                r = requests.post(url, json=json)
        elif request_type == "GET":
            response_logger.info("Sending GET request")
            r = requests.get(url)
    except requests.exceptions.ConnectionError as e:
        response_logger.error("Failed to connect to service", response="Connection error")
        raise e

    if 199 < r.status_code < 300:
        logger.info('Returned from service', response=r.reason, status_code=r.status_code)
    elif 399 < r.status_code < 500:
        logger.error('Returned from service', response=r.reason, status_code=r.status_code)
        raise ClientError(url=url, message=r.reason, status_code=r.status_code)
    elif r.status_code > 499:
        logger.error('Returned from service', response=r.reason, status_code=r.status_code)
        raise ServiceError(url=url, message=r.reason, status_code=r.status_code)

    return r


@app.route('/decrypt', methods=['POST', 'GET'])
@flask_security.login_required
def decrypt():
    audited_logger = logger.bind(user=flask_security.core.current_user.email)
    if request.method == "POST":
        data = request.form.get('EncryptedData')
        url = settings.SDX_DECRYPT_URL
        decrypted_data = ""

        try:
            audited_logger.info("Posting data to sdx-decrypt")
            decrypt_response = send_data(logger=audited_logger, url=url,
                                         data=data, request_type="POST")
        except ClientError:
            error = 'Client error'
        except ServiceError:
            error = 'Service error'
        except requests.exceptions.ConnectionError:
            error = 'Connection error'
        else:
            decrypted_data = decrypt_response.text
            error = ""

        return render_template('decrypt.html',
                               decrypted_data=decrypted_data,
                               error=error,
                               current_user=flask_security.core.current_user)

    else:
        return render_template('decrypt.html', current_user=flask_security.core.current_user)


def get_filtered_responses(logger, valid, tx_id, ru_ref, survey_id, datetime_earliest, datetime_latest):
    logger.info('Retrieving responses from sdx-store')
    try:
        q = db_session.query(SurveyResponse)
        if valid == "invalid":
            q = q.filter(SurveyResponse.invalid)
        elif valid == "valid":
            q = q.filter(not SurveyResponse.invalid)
        if tx_id != '':
            q = q.filter(SurveyResponse.tx_id == tx_id)
        if ru_ref != '':
            q = q.filter(SurveyResponse.data["metadata"]["ru_ref"].astext == ru_ref)
        if survey_id != '':
            q = q.filter(SurveyResponse.data["survey_id"].astext == survey_id)
        dt_column = func.date(SurveyResponse.data["submitted_at"].astext)
        if datetime_earliest:
            year = int(datetime_earliest[:4])
            month = int(datetime_earliest[5:7])
            day = int(datetime_earliest[8:10])
            hour = int(datetime_earliest[11:13])
            minute = int(datetime_earliest[14])
            q = q.filter(dt_column > datetime(year, month, day, hour, minute))
        if datetime_latest:
            year = int(datetime_latest[:4])
            month = int(datetime_latest[5:7])
            day = int(datetime_latest[8:10])
            hour = int(datetime_latest[11:13])
            minute = int(datetime_latest[14])
            q = q.filter(dt_column < datetime(year, month, day, hour, minute))
        filtered_data = q.all()
    except DataError as e:
        logger.error("Invalid search term", error=e)
        return []
    except SQLAlchemyError as e:
        logger.error("Database error", error=e)
        return []

    return filtered_data


def reprocess_transaction(logger, json_data):
    logger.info("Reprocessing transaction", tx_id=json_data["tx_id"])
    if json_data.get("invalid"):
        del json_data["invalid"]
    validate_response = send_data(
        logger=logger, url=settings.SDX_VALIDATE_URL, json=json_data, request_type="POST")
    if validate_response != 200:
        json_data['invalid'] = "True"
    send_data(logger=logger, url=settings.SDX_STORE_URL, json=json_data, request_type="POST")


@app.route('/store/', defaults={'page': 0}, methods=['GET', 'POST'])
@app.route('/store/<page>', methods=['GET', 'POST'])
@flask_security.login_required
def store(page):
    audited_logger = logger.bind(user=flask_security.core.current_user.email)
    if request.method == 'POST':
        json_survey_data = request.form.get('json_data')
        if not json_survey_data:
            json_survey_data = request.form.get('json_data_list')
            if not json_survey_data:
                return url_for('store')
            json_survey_data = ast.literal_eval(json_survey_data)

        if isinstance(json_survey_data, list):
            audited_logger.info("Reprocessing multiple transactions")
            for json_data in json_survey_data:
                reprocess_transaction(audited_logger, json_data)
        else:
            json_single_data = json.loads(json_survey_data.replace("'", '"'))
            reprocess_transaction(audited_logger, json_single_data)
        return redirect(url_for('store'))

    else:
        valid = request.args.get('valid', type=str, default='')
        tx_id = request.args.get('tx_id', type=str, default='')
        ru_ref = request.args.get('ru_ref', type=str, default='')
        survey_id = request.args.get('survey_id', type=str, default='')
        datetime_earliest = request.args.get('datetime_earliest', type=str, default='')
        datetime_latest = request.args.get('datetime_latest', type=str, default='')
        store_data = get_filtered_responses(
            audited_logger, valid, tx_id, ru_ref, survey_id, datetime_earliest, datetime_latest)
        audited_logger.info("Successfully retireved responses")
        json_list = [item.data for item in store_data]
        no_pages = math.ceil(round(float(len(json_list) / 20)))

        return render_template('store.html',
                               data=json_list,
                               no_pages=no_pages,
                               page=int(page),
                               current_user=flask_security.core.current_user)


@app.route('/storetest', methods=['GET'])
def storetest():

    def create_test_data(number):
        test_data = json.dumps(
            {
                "collection": {
                    "exercise_sid": "hfjdskf",
                    "instrument_id": "ce2016",
                    "period": "0616"
                },
                "data": {
                    "1": "2",
                    "2": "4",
                    "3": "2",
                    "4": "Y"
                },
                "metadata": {
                    "ru_ref": "12345678901a",
                    "user_id": "789473423"
                },
                "origin": "uk.gov.ons.edc.eq",
                "submitted_at": "2017-04-27T14:23:13+00:00",
                "survey_id": "1",
                "tx_id": "f088d89d-a367-876e-f29f-ae8f1a26" + str(number),
                "type": "uk.gov.ons.edc.eq:surveyresponse",
                "version": "0.0.1"
            })
        return test_data
    for i in range(1000, 1250):
        test_data = create_test_data(str(i))
        send_data(logger=logger, url=settings.SDX_STORE_URL, data=test_data, request_type="POST")

    return "data sent"


@app.route('/adduser', methods=['GET', 'POST'])
@flask_security.roles_required('Admin')
def add_user():
    form = NewUserForm()
    if request.method == 'POST' and form.validate():
        success = False
        try:
            create_dev_user(form.email.data, form.password.data)
            success = True
        except UserCreationError:
            form.errors['Database'] = ["Error creating user"]
        except UserExistsError:
            form.errors['User'] = ["This user already exists"]

        return render_template('adduser.html', form=form, success=success, user=form.email.data)

    else:
        return render_template('adduser.html', form=form)
